### PR TITLE
Process update rejection when workflow is completing itself

### PR DESCRIPTION
**Pull request by alexshtin**
## What changed?

Process update rejection and cancel incomplete updates when workflow is completing itself.

## Why?

Update rejection messages are processed after commands and if there is a command to complete workflow those messages were not processed at all. This PR fixes it.

Also all not accepted updates are canceled if workflow completes itself.

## How did you test it?

Added new unit and functional tests.

## Potential risks

No risks.

## Is hotfix candidate?

No.

### CODE REVIEW

File: service/history/workflow/util.go [hunk 1]
```diff
@@ -187,6 +190,9 @@ func TerminateWorkflow(
 		terminateIdentity,
 		deleteAfterTerminate,
 	)
+
+	// err = updateRegistry.CancelIncomplete(ctx, update.CancelReasonWorkflowTerminated, WithEffects(effects, mutableState))
```
Comment (alexshtin) [inline service/history/workflow/util.go:194]: I will uncomment this code in another PR.

File: service/history/workflow/mutable_state_impl.go [hunk 2]
```diff
@@ -3589,6 +3589,9 @@ func (ms *MutableStateImpl) ApplyWorkflowExecutionUpdateCompletedEvent(
 		}
 		sizeDelta = ui.Size() - sizeBefore
 	} else {
+		// TODO (alex): this should never happened because UpdateInfo should always be created before
```
Comment (ast2023) [inline service/history/workflow/mutable_state_impl.go:3592]: If something could only happen due to a bug, the best way is to panic. If this could happen due to a user/system error then you should return an error.

File: service/history/workflow/update/registry.go [hunk 3]
```diff
@@ -352,3 +343,18 @@ func (r *registry) findLocked(ctx context.Context, id string) (*Update, bool) {
 		withInstrumentation(&r.instrumentation),
 	), true
 }
+
+// enumerate returns a slice of all updates in the registry for which the
+// provided filter function returns true. The registry is locked for reading
+// while enumerating over the map. Resulted slice can be iterated w/o holding a lock.
+func (r *registry) enumerate(filter func(u *Update) bool) []*Update {
```
Comment (ast2023) [inline service/history/workflow/update/registry.go:350]: The standard name is filter for the function and predicate for the argument. Then you will not need to explain meaning in the docstring.

File: service/history/workflow/update/update.go [hunk 4]
```diff
@@ -289,6 +290,13 @@ func (u *Update) OnMessage(
 			return err
 		}
 	}
+
+	// If workflow was completed while processing this WFT, then only Rejection messages can be processed,
+	// because they don't create new events in the history. All other updates must be cancelled.
+	if _, isRejection := msg.(*updatepb.Rejection); !isWorkflowRunning && !isRejection {
```
Comment (ast2023) [inline service/history/workflow/update/update.go:296]: I think it will be easier to understand `!(isWorkflowRunning || isRejection)`, but even better approach is to define explanatory function or variable: `shouldCancel := !(isWorkflowRunning || isRejection)`

File: service/history/workflow/update/update.go [hunk 5]
```diff
@@ -494,6 +502,34 @@ func (u *Update) onResponseMsg(
...
+	return !u.state.Matches(stateSet(stateProvisionallyCompleted | stateCompleted))
+}
+
+// CancelIncomplete cancels update if it wasn't completed yet:
+//   - if in stateAdmitted, stateRequested, or stateSent -> reject,
+//   - if in stateAccepted -> do nothing,
+//   - if in stateCompleted -> do nothing.
+func (u *Update) CancelIncomplete(ctx context.Context, reason CancelReason, eventStore EventStore) error {
+	if u.state.Matches(stateSet(stateAdmitted | stateProvisionallyRequested | stateRequested | stateProvisionallySent | stateSent)) {
+		return u.reject(ctx, reason.RejectionFailure(), eventStore)
```
Comment (ast2023) [inline service/history/workflow/update/update.go:516]: What I love about comments, is that they frequently lye. Docstring for this function does not match the functionality.

File: service/history/workflow/update/registry.go [hunk 6]
```diff
@@ -195,10 +196,18 @@ func (r *registry) Find(ctx context.Context, id string) (*Update, bool) {
...
-func (r *registry) TerminateUpdates(_ context.Context, _ EventStore) {
-	// TODO (alex-update): implement
-	// This method is not implemented and update API callers will just timeout.
-	// In future, it should remove all existing updates and notify callers with better error.
+// CancelIncomplete cancels all incomplete updates in the registry:
+//   - updates in stateAdmitted, stateRequested, or stateSent are rejected,
+//   - updates in stateAccepted are ignored (see CancelIncomplete() in update.go for details),
+//   - updates in stateCompleted are ignored.
+func (r *registry) CancelIncomplete(ctx context.Context, reason CancelReason, eventStore EventStore) error {
+	incompleteUpdates := r.enumerate(func(u *Update) bool { return u.isIncomplete() })
```
Comment (ast2023) [inline service/history/workflow/update/registry.go:204]: It looks like you filter all the updates which are complete. What is the reason to call them `incompleUpdates`

File: service/history/workflow/update/update_test.go [hunk 7]
```diff
@@ -904,6 +905,81 @@ func TestWaitLifecycleStage(t *testing.T) {
...
+		ctx     = context.Background()
+		effects = effect.Immediate(ctx)
+
+		meta = updatepb.Meta{UpdateId: t.Name() + "-update-id"}
+		req  = updatepb.Request{Meta: &meta, Input: &updatepb.Input{Name: t.Name()}}
+
+		store = mockEventStore{Controller: effects}
+	)
+
+	t.Run("request", func(t *testing.T) {
```
Comment (ast2023) [inline service/history/workflow/update/update_test.go:919]: this is more then "request". Could you clarify?

File: service/history/workflow/update/update_test.go [hunk 8]
```diff
@@ -904,6 +905,81 @@ func TestWaitLifecycleStage(t *testing.T) {
...
+		require.NoError(t, err)
+
+		status, err := upd.WaitOutcome(ctx)
+		require.NoError(t, err)
+		require.Equal(t, enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_COMPLETED, status.Stage)
+		require.Equal(t, "Workflow Update is rejected because Workflow Execution is completed.", status.Outcome.GetFailure().GetMessage())
+		require.Equal(t, "CanceledUpdate", status.Outcome.GetFailure().GetApplicationFailureInfo().Type)
+	})
+
+	t.Run("accept", func(t *testing.T) {
```
Comment (ast2023) [inline service/history/workflow/update/update_test.go:931]: This test and the previous differ in the third parameter of OnMessage. In this case the workflow is running. I do not understand why we should get failure and the test name does not explain it.

File: service/history/workflow_task_handler.go [hunk 9]
```diff
@@ -351,7 +355,8 @@ func (handler *workflowTaskHandlerImpl) handleMessage(
 				enumspb.WORKFLOW_TASK_FAILED_CAUSE_BAD_UPDATE_WORKFLOW_EXECUTION_MESSAGE,
 				serviceerror.NewNotFound(fmt.Sprintf("update %q not found", message.ProtocolInstanceId)))
 		}
-		if err := upd.OnMessage(ctx, message, workflow.WithEffects(handler.effects, handler.mutableState)); err != nil {
+
+		if err := upd.OnMessage(ctx, message, handler.mutableState.IsWorkflowExecutionRunning(), workflow.WithEffects(handler.effects, handler.mutableState)); err != nil {
```
Comment (ast2023) [inline service/history/workflow_task_handler.go:359]: Do we have limit on number of characters in line? Anyway, people read faster at around 80 and common agreement in code is around 120. I've heard that it is impolite to write lines which make people to scroll right to read.

File: service/history/workflow/update/registry.go (hunk 6 above)
Comment (alexshtin) [reply service/history/workflow/update/registry.go:204]: `isIncomplete` returns true if update is not complete.

File: service/history/workflow_task_handler_callbacks.go [hunk 10]
```diff
@@ -606,6 +606,23 @@ func (handler *workflowTaskHandlerCallbacksImpl) handleWorkflowTaskCompleted(
 			return nil, err
 		}
 
+		if !ms.IsWorkflowExecutionRunning() {
```
Comment (ast2023) [inline service/history/workflow_task_handler_callbacks.go:609]: I suggest we extract this code into a separate method. This method is too long already.

File: tests/update_workflow_test.go [hunk 11]
```diff
@@ -3290,35 +3290,54 @@ func (s *FunctionalSuite) TestUpdateWorkflow_ScheduledSpeculativeWorkflowTask_Te
...
+			UpdateFailure: "Workflow Update is rejected because Workflow Execution is completed.",
+			Commands:      func(_ *testvars.TestVars) []*commandpb.Command { return nil },
+			Messages:      func(_ *testvars.TestVars, _ *protocolpb.Message) []*protocolpb.Message { return nil },
 		},
 		{
-			Name:         "accepted",
-			UpdateErrMsg: "update has been terminated",
-			Commands:     func(tv *testvars.TestVars) []*commandpb.Command { return s.acceptUpdateCommands(tv, "1") },
+			Name:          "accepted",
+			Description:   "update in stateAccepted must got an error because there is no way for workflow to complete it and already accepted update can't be rejected",
```
Comment (ast2023) [inline tests/update_workflow_test.go:3313]: there is misspelling, got vs get

File: service/history/workflow/update/update.go [hunk 12]
```diff
@@ -276,6 +276,7 @@ func (u *Update) WaitAccepted(ctx context.Context) (UpdateStatus, error) {
 func (u *Update) OnMessage(
 	ctx context.Context,
 	msg proto.Message,
+	isWorkflowRunning bool,
```
Comment (ast2023) [inline service/history/workflow/update/update.go:279]: Comments usually out of sync with the code because we update code but not the comments. It will be better to define a new type message which consist of two fields proto.Message and isWorkflowRunning. Then move onUpdate parsing of the message part to the new message class. This will leave OnMessage a pure double dispatch.

File: service/history/workflow/update/registry.go (hunk 6 above)
Comment (ast2023) [reply service/history/workflow/update/registry.go:204]: Ah, yes, thank you. Probably too tired to review code.

File: service/history/workflow/mutable_state_impl.go (hunk 2 above)
Comment (alexshtin) [reply service/history/workflow/mutable_state_impl.go:3592]: I will think about it when address this TODO. But all panics on the API call code path (like this one) are captured and converted to internal errors which is returned to the client. Returning specific error makes more sense here.

File: service/history/workflow/update/update.go (hunk 4 above)
Comment (alexshtin) [reply service/history/workflow/update/update.go:296]: I tried to leverage go `if` one liner syntax.

File: service/history/workflow/update/update.go (hunk 5 above)
Comment (alexshtin) [reply service/history/workflow/update/update.go:516]: Do you mean that I didn't includes Provisional states in docstring? This is on purpose. It is not not important detail and I want docstring to explain what this method does, not to rewrite in plain English.

File: service/history/workflow_task_handler_callbacks.go [hunk 13]
```diff
@@ -606,6 +606,23 @@ func (handler *workflowTaskHandlerCallbacksImpl) handleWorkflowTaskCompleted(
...
+			// If workflow competed itself with one of the completion command, cancel all incomplete updates in the registry.
+			// Because all unprocessed updates were already rejected, incomplete updates in the registry are:
+			// - updates that were received while this WT was running,
+			// - updates that were accepted but not completed by this WT.
+			err = weContext.UpdateRegistry(ctx).CancelIncomplete(ctx, update.CancelReasonWorkflowCompleted, workflow.WithEffects(&effects, ms))
```
Comment (yycptt) [inline service/history/workflow_task_handler_callbacks.go:614]: Are updates rejected before or after we persist the mutable state change? If it's before, then the persistence could fail, and workflow remain open while updates are already rejected due to workflow completed.

File: service/history/api/updateworkflow/api.go [hunk 14]
```diff
@@ -125,7 +125,7 @@ func Invoke(
 			if upd, alreadyExisted, err = updateReg.FindOrCreate(ctx, updateID); err != nil {
 				return nil, err
 			}
-			if err = upd.OnMessage(ctx, req.GetRequest().GetRequest(), workflow.WithEffects(effect.Immediate(ctx), ms)); err != nil {
+			if err = upd.OnMessage(ctx, req.GetRequest().GetRequest(), ms.IsWorkflowExecutionRunning(), workflow.WithEffects(effect.Immediate(ctx), ms)); err != nil {
```
Comment (yycptt) [inline service/history/api/updateworkflow/api.go:128]: If when the update request is received the workflow is already closed, we will return NotFound error (on L96 above), but if workflow is closed after update is received but before it's sent, we will reject the update with some message. I think it will be ideal to make them consistent.

File: service/history/workflow/update/registry.go [hunk 15]
```diff
@@ -195,10 +196,18 @@ func (r *registry) Find(ctx context.Context, id string) (*Update, bool) {
...
 
-func (r *registry) TerminateUpdates(_ context.Context, _ EventStore) {
-	// TODO (alex-update): implement
-	// This method is not implemented and update API callers will just timeout.
-	// In future, it should remove all existing updates and notify callers with better error.
+// CancelIncomplete cancels all incomplete updates in the registry:
+//   - updates in stateAdmitted, stateRequested, or stateSent are rejected,
+//   - updates in stateAccepted are ignored (see CancelIncomplete() in update.go for details),
+//   - updates in stateCompleted are ignored.
+func (r *registry) CancelIncomplete(ctx context.Context, reason CancelReason, eventStore EventStore) error {
```
Comment (yycptt) [inline service/history/workflow/update/registry.go:203]: I wonder if there's some central place that allow us to perform this operation on workflow close, so we don't have to remember to call it on every path that lead to workflow close.

File: service/history/workflow_task_handler_callbacks.go (hunk 13 above)
Comment (alexshtin) [reply service/history/workflow_task_handler_callbacks.go:614]: All operations on futures are buffered and applied only after MS is persisted. `effects` (last parameter) is responsible for it.

File: service/history/workflow/update/registry.go (hunk 15 above)
Comment (alexshtin) [reply service/history/workflow/update/registry.go:203]: Moreover, there is no single place even for all cases where we terminate WF. This why I postponed these changes to another PR. And there is also WF timeout case. I will create a task to refactor this.

File: service/history/workflow/update/registry_test.go [hunk 16]
```diff
@@ -568,3 +570,85 @@ func TestRejectUnprocessed(t *testing.T) {
 	require.NoError(t, err)
 	require.Len(t, rejectedIDs, 0, "rejected updates shouldn't be rejected again")
 }
+
+func TestCancelIncomplete(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (alexshtin) [reply service/history/workflow/update/registry_test.go:574]: Too much of boilerplate code? It creates updates in different states, calls `CancelIncomplete` and check the state of every update after that. I will rename update variables to reflect this.

File: service/history/workflow/update/update_test.go (hunk 7 above)
Comment (alexshtin) [reply service/history/workflow/update/update_test.go:919]: Gave them a better name.

File: service/history/workflow_task_handler.go (hunk 9 above)
Comment (alexshtin) [reply service/history/workflow_task_handler.go:359]: We don't have any rules on this and I personally try not to cut lines if it is not necessary. I am happy to adopt any rule. Do you think that 120 is some sort of standart?

File: service/history/workflow_task_handler_callbacks.go (hunk 10 above)
Comment (alexshtin) [reply service/history/workflow_task_handler_callbacks.go:609]: It is a call to a single function + error handling. This entire function must be split in may sub function. I will do it later.

File: tests/update_workflow_test.go (hunk 11 above)
Comment (alexshtin) [reply tests/update_workflow_test.go:3313]: sharp eye!

File: service/history/workflow/update/registry_test.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (ast2023) [reply service/history/workflow/update/registry_test.go:574]: It would be nice to see each test case as a separate test, just like you did in TestCompletedWorkflow.

File: service/history/api/updateworkflow/api.go (hunk 14 above)
Comment (yycptt) [reply service/history/api/updateworkflow/api.go:128]: yeah that SGTM.

File: service/history/workflow/update/registry.go (hunk 15 above)
Comment (yycptt) [reply service/history/workflow/update/registry.go:203]: hmm if mutable state has access to the registry then probably we can perform the operation there. 🤔

File: service/history/workflow_task_handler_callbacks.go (hunk 13 above)
Comment (yycptt) [reply service/history/workflow_task_handler_callbacks.go:614]: ah, I see. I am still not familiar with the effect package enough.

**Review by yycptt — APPROVED** [review]

**Review by ast2023 — APPROVED** [review]

File: service/history/workflow/update/update.go [hunk 17]
```diff
@@ -494,6 +504,34 @@ func (u *Update) onResponseMsg(
 	return nil
 }
 
+// isIncomplete checks if update is already completed (rejected or processed).
+func (u *Update) isIncomplete() bool {
+	return !u.state.Matches(stateSet(stateProvisionallyCompleted | stateCompleted))
```
Comment (dandavison) [inline service/history/workflow/update/update.go:509]: @alexshtin was there a clear reason for including `stateProvisionallyCompleted` here? It can be rolled-back (is not durable) so the update is not necessarily complete. The reason I'm asking is that I need to skip reapply in the case where an update is in-progress, and I could use the `isIncomplete` method as the definition of in-progress. But I think that I would want to define `stateProvisionallyCompleted` as in-progress for that use case.
